active_discussion/validity.md

```diff
@@ -0,0 +1,131 @@
+# Data type validity requirements
+
+This discussion is meant to focus on the question: Which invariants derived from
+types are there that the compiler expects to be *always* maintained, and
+(equivalently) that unsafe code must *always* uphold (or else cause undefined
+behavior)?  This is what is called "validity invariant" in
```

Comment:
I'm curious whether there exist properties that unsafe code must always uphold but which it cannot **rely upon**. The idea would be: for now, we require that this is true, but we expect we may not require it forever. Technically, from the compiler's POV, we can tweak these predicates because that just makes less code UB -- but unsafe code could be relying on the fact that doing otherwise would be UB, and that could be a problem.

This seems like a more general thing -- it also e.g. arises in the reprsentation discussion -- but once we start getting into invariants, I expect we should think carefully about this distinction.

I would sort of like this to be noted somewhere as a kind of "cross-cutting concern" -- i.e., for each area, we are aiming to define these two predicates (what unsafe code must uphold, and what it can rely on) somewhat separately.

Reply:
I agree in general.  We want some "wiggle room" in our definition of what unsafe code can do, certainly when it comes to the safety invariant.  I am less sure for the validity invariant.

My main problem with this is that I don't have a good understanding yet of how this works formally.  That's mostly because I did not spend enough time to think about this.